docs/manpages/rpcclient.1
.\" This manpage has been automatically generated by docbook2man-spec
.\" from a DocBook document.  docbook2man-spec can be found at:
.\" <http://shell.ipoline.com/~elmert/hacks/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "RPCCLIENT" "1" "28 March 2001" "" ""
.SH NAME
rpcclient \- tool for executing client side  MS-RPC functions
.SH SYNOPSIS
.sp
\fBrpcclient\fR \fBserver\fR [ \fB-A authfile\fR ]  [ \fB-c <command string>\fR ]  [ \fB-d debuglevel\fR ]  [ \fB-h\fR ]  [ \fB-l logfile\fR ]  [ \fB-N\fR ]  [ \fB-s <smb config file>\fR ]  [ \fB-U username[%password]\fR ]  [ \fB-W workgroup\fR ]  [ \fB-N\fR ] 
.SH "DESCRIPTION"
.PP
This tool is part of the  Sambasuite.
.PP
\fBrpcclient\fR is a utility initially developed
to test MS-RPC functionality in Samba itself. It has undergone 
several stages of development and stability. Many system administrators
have now written scripts around it to manage Windows NT clients from 
their UNIX workstation. 
.SH "OPTIONS"
.TP
\fBserver\fR
NetBIOS name of Server to which to connect. 
The server can be any SMB/CIFS server. The name is 
resolved using the  \fIname resolve order\fRline from 
\fIsmb.conf(5)\fR.
.TP
\fB-A filename\fR
This option allows 
you to specify a file from which to read the username and 
password used in the connection. The format of the file is 

.sp
.nf
		username = <value> 
		password = <value>
		domain   = <value>
		
.sp
.fi

Make certain that the permissions on the file restrict 
access from unwanted users. 
.TP
\fB-c 'command string'\fR
execute semicolon separated commands (listed 
below)) 
.TP
\fB-d debuglevel\fR
set the debuglevel. Debug level 0 is the lowest 
and 100 being the highest. This should be set to 100 if you are
planning on submitting a bug report to the Samba team (see BUGS.txt). 
.TP
\fB-h\fR
Print a summary of command line options.
.TP
\fB-l logbasename\fR
File name for log/debug files. .client will be 
appended. The log file is never removed by the client.
.TP
\fB-N\fR
instruct \fBrpcclient\fR not to ask 
for a password. By default, \fBrpcclient\fR will prompt 
for a password. See also the \fI-U\fR option.
.TP
\fB-s smb.conf\fR
Specifies the location of the all important 
\fIsmb.conf\fR file. 
.TP
\fB-U username[%password]\fR
Sets the SMB username or username and password. 

If %password is not specified, The user will be prompted. The 
client will first check the \fBUSER\fR environment variable, then the 
\fBLOGNAME\fR variable and if either exist, the 
string is uppercased. If these environmental variables are not 
found, the username GUEST is used. 

A third option is to use a credentials file which 
contains the plaintext of the username and password. This 
option is mainly provided for scripts where the admin doesn't 
desire to pass the credentials on the command line or via environment 
variables. If this method is used, make certain that the permissions 
on the file restrict access from unwanted users. See the 
\fI-A\fR for more details. 

Be cautious about including passwords in scripts. Also, on 
many systems the command line of a running process may be seen 
via the \fBps\fR command. To be safe always allow 
\fBrpcclient\fR to prompt for a password and type 
it in directly. 
.TP
\fB-W domain\fR
Set the SMB domain of the username. This 
overrides the default domain which is the domain of the 
server specified with the \fI-S\fR option. 
If the domain specified is the same as the server's NetBIOS name, 
it causes the client to log on using the server's local SAM (as 
opposed to the Domain SAM). 
.SH "COMMANDS"
.PP
\fBLSARPC\fR
.TP 0.2i
\(bu
\fBlsaquery\fR
.TP 0.2i
\(bu
\fBlookupsids\fR
.TP 0.2i
\(bu
\fBlookupnames\fR
.TP 0.2i
\(bu
\fBenumtrusts\fR
.PP
.PP
.PP
\fBSAMR\fR
.PP
.TP 0.2i
\(bu
\fBqueryuser\fR
.TP 0.2i
\(bu
\fBquerygroup\fR
.TP 0.2i
\(bu
\fBqueryusergroups\fR
.TP 0.2i
\(bu
\fBquerygroupmem\fR
.PP
.PP
.PP
\fBSPOOLSS\fR
.PP
.TP 0.2i
\(bu
\fBadddriver <arch> <config>\fR 
- Execute an AddPrinterDriver() RPC to install the printer driver 
information on the server. Note that the driver files should 
already exist in the directory returned by 
\fBgetdriverdir\fR. Possible values for 
\fIarch\fR are the same as those for 
the \fBgetdriverdir\fR command.
The \fIconfig\fR parameter is defined as 
follows: 

.sp
.nf
		Long Printer Name:\\
		Driver File Name:\\
		Data File Name:\\
		Config File Name:\\
		Help File Name:\\
		Language Monitor Name:\\
		Default Data Type:\\
		Comma Separated list of Files
		
.sp
.fi

Any empty fields should be enter as the string "NULL". 

Samba does not need to support the concept of Print Monitors
since these only apply to local printers whose driver can make
use of a bi-directional link for communication. This field should 
be "NULL". On a remote NT print server, the Print Monitor for a 
driver must already be installed prior to adding the driver or 
else the RPC will fail. 
.TP 0.2i
\(bu
\fBaddprinter <printername> 
<sharename> <drivername> <port>\fR 
- Add a printer on the remote server. This printer 
will be automatically shared. Be aware that the printer driver 
must already be installed on the server (see \fBadddriver\fR) 
and the \fIport\fRmust be a valid port name (see
\fBenumports\fR.
.TP 0.2i
\(bu
\fBenumdata\fR - Enumerate all 
printer setting data stored on the server. On Windows NT clients, 
these values are stored in the registry, while Samba servers 
store them in the printers TDB. This command corresponds
to the MS Platform SDK GetPrinterData() function (* This
command is currently unimplemented).
.TP 0.2i
\(bu
\fBenumjobs <printer>\fR 
- List the jobs and status of a given printer. 
This command corresponds to the MS Platform SDK EnumJobs() 
function (* This command is currently unimplemented).
.TP 0.2i
\(bu
\fBenumports [level]\fR 
- Executes an EnumPorts() call using the specified 
info level. Currently only info levels 1 and 2 are supported. 
.TP 0.2i
\(bu
\fBenumdrivers [level]\fR 
- Execute an EnumPrinterDrivers() call. This lists the various installed 
printer drivers for all architectures. Refer to the MS Platform SDK 
documentation for more details of the various flags and calling 
options. Currently supported info levels are 1, 2, and 3.
.TP 0.2i
\(bu
\fBenumprinters [level]\fR 
- Execute an EnumPrinters() call. This lists the various installed 
and share printers. Refer to the MS Platform SDK documentation for 
more details of the various flags and calling options. Currently
supported info levels are 0, 1, and 2.
.TP 0.2i
\(bu
\fBgetdata <printername>\fR 
- Retrieve the data for a given printer setting. See 
the \fBenumdata\fR command for more information. 
This command corresponds to the GetPrinterData() MS Platform 
SDK function (* This command is currently unimplemented). 
.TP 0.2i
\(bu
\fBgetdriver <printername>\fR 
- Retrieve the printer driver information (such as driver file, 
config file, dependent files, etc...) for 
the given printer. This command corresponds to the GetPrinterDriver()
MS Platform SDK function. Currently info level 1, 2, and 3 are supported.
.TP 0.2i
\(bu
\fBgetdriverdir <arch>\fR 
- Execute a GetPrinterDriverDirectory()
RPC to retreive the SMB share name and subdirectory for 
storing printer driver files for a given architecture. Possible 
values for \fIarch\fR are "Windows 4.0" 
(for Windows 95/98), "Windows NT x86", "Windows NT PowerPC", "Windows
Alpha_AXP", and "Windows NT R4000". 
.TP 0.2i
\(bu
\fBgetprinter <printername>\fR 
- Retrieve the current printer information. This command 
corresponds to the GetPrinter() MS Platform SDK function. 
.TP 0.2i
\(bu
\fBopenprinter <printername>\fR 
- Execute an OpenPrinterEx() and ClosePrinter() RPC 
against a given printer. 
.TP 0.2i
\(bu
\fBsetdriver <printername> <drivername>\fR 
- Execute a SetPrinter() command to update the printer driver associated
with an installed printer. The printer driver must already be correctly
installed on the print server. 

See also the \fBenumprinters\fR and 
\fBenumdrivers\fR commands for obtaining a list of
of installed printers and drivers.
.PP
\fBGENERAL OPTIONS\fR
.PP
.TP 0.2i
\(bu
\fBdebuglevel\fR - Set the current debug level
used to log information.
.TP 0.2i
\(bu
\fBhelp (?)\fR - Print a listing of all 
known commands or extended help on a particular command. 
.TP 0.2i
\(bu
\fBquit (exit)\fR - Exit \fBrpcclient
\fR\&.
.SH "BUGS"
.PP
\fBrpcclient\fR is designed as a developer testing tool 
and may not be robust in certain areas (such as command line parsing). 
It has been known to generate a core dump upon failures when invalid 
parameters where passed to the interpreter. 
.PP
From Luke Leighton's original rpcclient man page:
.PP
\fB"WARNING!\fR The MSRPC over SMB code has 
been developed from examining Network traces. No documentation is 
available from the original creators (Microsoft) on how MSRPC over 
SMB works, or how the individual MSRPC services work. Microsoft's 
implementation of these services has been demonstrated (and reported) 
to be... a bit flakey in places. 
.PP
The development of Samba's implementation is also a bit rough, 
and as more of the services are understood, it can even result in 
versions of \fBsmbd(8)\fR and \fBrpcclient(1)\fR 
that are incompatible for some commands or services. Additionally, 
the developers are sending reports to Microsoft, and problems found 
or reported to Microsoft are fixed in Service Packs, which may 
result in incompatibilities." 
.SH "VERSION"
.PP
This man page is correct for version 2.2 of the Samba 
suite.
.SH "AUTHOR"
.PP
The original Samba software and related utilities 
were created by Andrew Tridgell. Samba is now developed
by the Samba Team as an Open Source project similar 
to the way the Linux kernel is developed.
.PP
The original rpcclient man page was written by Matthew 
Geddes, Luke Kenneth Casson, and rewriten by Gerald Carter. 
The conversion to DocBook for Samba 2.2 was done by Gerald 
Carter.


NEWS
Changes in release 0.3d

 * add krb5-config

 * fix a bug in 3des gss-api mechanism, making it compatible with the
   specification and the MIT implementation

 * make telnetd only allow a specific list of environment variables to
   stop it from setting `sensitive' variables

 * try to use an existing libdes

 * lib/krb5, kdc: use correct usage type for ap-req messages.  This
   should improve compatability with MIT krb5 when using 3DES
   encryption types

 * kdc: fix memory allocation problem

 * update config.guess and config.sub

 * lib/roken: more stuff implemented

 * bug fixes and portability enhancements

Changes in release 0.3c

 * lib/krb5: memory caches now support the resolve operation

 * appl/login: set PATH to some sane default

 * kadmind: handle several realms

 * bug fixes (including memory leaks)

Changes in release 0.3b

 * kdc: prefer default-salted keys on v5 requests

 * kdc: lowercase hostnames in v4 mode

 * hprop: handle more types of MIT salts

 * lib/krb5: fix memory leak

 * bug fixes

Changes in release 0.3a:

 * implement arcfour-hmac-md5 to interoperate with W2K

 * modularise the handling of the master key, and allow for other
   encryption types. This makes it easier to import a database from
   some other source without having to re-encrypt all keys.

 * allow for better control over which encryption types are created

 * make kinit fallback to v4 if given a v4 KDC

 * make klist work better with v4 and v5, and add some more MIT
   compatibility options

 * make the kdc listen on the krb524 (4444) port for compatibility
   with MIT krb5 clients

 * implement more DCE/DFS support, enabled with --enable-dce, see
   lib/kdfs and appl/dceutils

 * make the sequence numbers work correctly

 * bug fixes

Changes in release 0.2t:

 * bug fixes

Changes in release 0.2s:

 * add OpenLDAP support in hdb

 * login will get v4 tickets when it receives forwarded tickets

 * xnlock supports both v5 and v4

 * repair source routing for telnet

 * fix building problems with krb4 (krb_mk_req)

 * bug fixes

Changes in release 0.2r:

 * fix realloc memory corruption bug in kdc

 * `add --key' and `cpw --key' in kadmin

 * klist supports listing v4 tickets

 * update config.guess and config.sub

 * make v4 -> v5 principal name conversion more robust

 * support for anonymous tickets

 * new man-pages

 * telnetd: do not negotiate KERBEROS5 authentication if there's no keytab.

 * use and set expiration and not password expiration when dumping
   to/from ka server databases / krb4 databases

 * make the code happier with 64-bit time_t

 * follow RFC2782 and by default do not look for non-underscore SRV names

Changes in release 0.2q:

 * bug fix in tcp-handling in kdc

 * bug fix in expand_hostname

Changes in release 0.2p:

 * bug fix in `kadmin load/merge'

 * bug fix in krb5_parse_address

Changes in release 0.2o:

 * gss_{import,export}_sec_context added to libgssapi

 * new option --addresses to kdc (for listening on an explicit set of
   addresses)

 * bug fixes in the krb4 and kaserver emulation part of the kdc

 * other bug fixes

Changes in release 0.2n:

 * more robust parsing of dump files in kadmin
 * changed default timestamp format for log messages to extended ISO
   8601 format (Y-M-DTH:M:S)
 * changed md4/md5/sha1 APIes to be de-facto `standard'
 * always make hostname into lower-case before creating principal
 * small bits of more MIT-compatability
 * bug fixes

Changes in release 0.2m:

 * handle glibc's getaddrinfo() that returns several ai_canonname

 * new endian test

 * man pages fixes

Changes in release 0.2l:

 * bug fixes

Changes in release 0.2k:

 * better IPv6 test

 * make struct sockaddr_storage in roken work better on alphas

 * some missing [hn]to[hn]s fixed.

 * allow users to change their own passwords with kadmin (with initial
   tickets)

 * fix stupid bug in parsing KDC specification

 * add `ktutil change' and `ktutil purge'

Changes in release 0.2j:

 * builds on Irix

 * ftpd works in passive mode

 * should build on cygwin

 * work around broken IPv6-code on OpenBSD 2.6, also add configure
   option --disable-ipv6

Changes in release 0.2i:

 * use getaddrinfo in the missing places.

 * fix SRV lookup for admin server

 * use get{addr,name}info everywhere.  and implement it in terms of
   getipnodeby{name,addr} (which uses gethostbyname{,2} and
   gethostbyaddr)

Changes in release 0.2h:

 * fix typo in kx (now compiles)

Changes in release 0.2g:

 * lots of bug fixes:
   * push works
   * repair appl/test programs
   * sockaddr_storage works on solaris (alignment issues)
   * works better with non-roken getaddrinfo
   * rsh works
   * some non standard C constructs removed

Changes in release 0.2f:

 * support SRV records for kpasswd
 * look for both _kerberos and krb5-realm when doing host -> realm mapping

Changes in release 0.2e:

 * changed copyright notices to remove `advertising'-clause.
 * get{addr,name}info added to roken and used in the other code
   (this makes things work much better with hosts with both v4 and v6
    addresses, among other things)
 * do pre-auth for both password and key-based get_in_tkt
 * support for having several databases
 * new command `del_enctype' in kadmin
 * strptime (and new strftime) add to roken
 * more paranoia about finding libdb
 * bug fixes

Changes in release 0.2d:

 * new configuration option [libdefaults]default_etypes_des
 * internal ls in ftpd builds without KRB4
 * kx/rsh/push/pop_debug tries v5 and v4 consistenly
 * build bug fixes
 * other bug fixes

Changes in release 0.2c:

 * bug fixes (see ChangeLog's for details)

Changes in release 0.2b:

 * bug fixes
 * actually bump shared library versions

Changes in release 0.2a:

 * a new program verify_krb5_conf for checking your /etc/krb5.conf
 * add 3DES keys when changing password
 * support null keys in database
 * support multiple local realms
 * implement a keytab backend for AFS KeyFile's

source4/rpcclient/rpcclient.c
/* 
   Unix SMB/CIFS implementation.
   RPC pipe client

   Copyright (C) Tim Potter 2000-2001
   Copyright (C) Martin Pool 2003

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "includes.h"
#include "rpcclient.h"

DOM_SID domain_sid;


/* List to hold groups of commands.
 *
 * Commands are defined in a list of arrays: arrays are easy to
 * statically declare, and lists are easier to dynamically extend.
 */

static struct cmd_list {
	struct cmd_list *prev, *next;
	struct cmd_set *cmd_set;
} *cmd_list;

/****************************************************************************
handle completion of commands for readline
****************************************************************************/
static char **completion_fn(char *text, int start, int end)
{
#define MAX_COMPLETIONS 100
	char **matches;
	int i, count=0;
	struct cmd_list *commands = cmd_list;

#if 0	/* JERRY */
	/* FIXME!!!  -- what to do when completing argument? */
	/* for words not at the start of the line fallback 
	   to filename completion */
	if (start) 
		return NULL;
#endif

	/* make sure we have a list of valid commands */
	if (!commands) 
		return NULL;

	matches = (char **)malloc(sizeof(matches[0])*MAX_COMPLETIONS);
	if (!matches) return NULL;

	matches[count++] = strdup(text);
	if (!matches[0]) return NULL;

	while (commands && count < MAX_COMPLETIONS-1) 
	{
		if (!commands->cmd_set)
			break;
		
		for (i=0; commands->cmd_set[i].name; i++)
		{
			if ((strncmp(text, commands->cmd_set[i].name, strlen(text)) == 0) &&
				commands->cmd_set[i].fn) 
			{
				matches[count] = strdup(commands->cmd_set[i].name);
				if (!matches[count]) 
					return NULL;
				count++;
			}
		}
		
		commands = commands->next;
		
	}

	if (count == 2) {
		SAFE_FREE(matches[0]);
		matches[0] = strdup(matches[1]);
	}
	matches[count] = NULL;
	return matches;
}

/***********************************************************************
 * read in username/password credentials from a file
 */
static void read_authfile (
	char *filename, 
	char* username, 
	char* password, 
	char* domain
)
{
	FILE *auth;
        fstring buf;
        uint16 len = 0;
	char *ptr, *val, *param;
                               
	if ((auth=sys_fopen(filename, "r")) == NULL)
	{
		printf ("ERROR: Unable to open credentials file!\n");
		return;
	}
                                
	while (!feof(auth))
	{  
		/* get a line from the file */
		if (!fgets (buf, sizeof(buf), auth))
			continue;
		
		len = strlen(buf);
		
		/* skip empty lines */			
		if ((len) && (buf[len-1]=='\n'))
		{
			buf[len-1] = '\0';
			len--;
		}	
		if (len == 0)
			continue;
					
		/* break up the line into parameter & value.
		   will need to eat a little whitespace possibly */
		param = buf;
		if (!(ptr = strchr_m(buf, '=')))
			continue;
		val = ptr+1;
		*ptr = '\0';
					
		/* eat leading white space */
		while ((*val!='\0') && ((*val==' ') || (*val=='\t')))
			val++;
					
		if (strwicmp("password", param) == 0)
			fstrcpy (password, val);
		else if (strwicmp("username", param) == 0)
			fstrcpy (username, val);
		else if (strwicmp("domain", param) == 0)
			fstrcpy (domain, val);
						
		memset(buf, 0, sizeof(buf));
	}
	fclose(auth);
	
	return;
}

static char* next_command (char** cmdstr)
{
	static pstring 		command;
	char			*p;
	
	if (!cmdstr || !(*cmdstr))
		return NULL;
	
	p = strchr_m(*cmdstr, ';');
	if (p)
		*p = '\0';
	pstrcpy(command, *cmdstr);
	if (p)
		*cmdstr = p + 1;
	else
		*cmdstr = NULL;
	
	return command;
}


/**
 * Find default username from environment variables.
 *
 * @param username fstring to receive username; not touched if none is
 * known.
 **/
static void get_username (char *username)
{
        if (getenv("USER"))
                fstrcpy(username,getenv("USER"));
 
        if (*username == 0 && getenv("LOGNAME"))
                fstrcpy(username,getenv("LOGNAME"));
 
        if (*username == 0) {
                fstrcpy(username,"GUEST");
        }

	return;
}

/* Fetch the SID for this computer */

static void fetch_machine_sid(struct cli_state *cli)
{
	POLICY_HND pol;
	NTSTATUS result = NT_STATUS_OK;
	uint32 info_class = 5;
	fstring domain_name;
	static BOOL got_domain_sid;
	TALLOC_CTX *mem_ctx;

	if (got_domain_sid) return;

	if (!(mem_ctx=talloc_init("fetch_machine_sid")))
	{
		DEBUG(0,("fetch_machine_sid: talloc_init returned NULL!\n"));
		goto error;
	}


	if (!cli_nt_session_open (cli, PI_LSARPC)) {
		fprintf(stderr, "could not initialise lsa pipe\n");
		goto error;
	}
	
	result = cli_lsa_open_policy(cli, mem_ctx, True, 
				     SEC_RIGHTS_MAXIMUM_ALLOWED,
				     &pol);
	if (!NT_STATUS_IS_OK(result)) {
		goto error;
	}

	result = cli_lsa_query_info_policy(cli, mem_ctx, &pol, info_class, 
					   domain_name, &domain_sid);
	if (!NT_STATUS_IS_OK(result)) {
		goto error;
	}

	got_domain_sid = True;

	cli_lsa_close(cli, mem_ctx, &pol);
	cli_nt_session_close(cli);
	talloc_destroy(mem_ctx);

	return;

 error:
	fprintf(stderr, "could not obtain sid for domain %s\n", cli->domain);

	if (!NT_STATUS_IS_OK(result)) {
		fprintf(stderr, "error: %s\n", nt_errstr(result));
	}

	exit(1);
}

/* List the available commands on a given pipe */

static NTSTATUS cmd_listcommands(struct cli_state *cli, TALLOC_CTX *mem_ctx,
				 int argc, const char **argv)
{
	struct cmd_list *tmp;
        struct cmd_set *tmp_set;
	int i;

        /* Usage */

        if (argc != 2) {
                printf("Usage: %s <pipe>\n", argv[0]);
                return NT_STATUS_OK;
        }

        /* Help on one command */

	for (tmp = cmd_list; tmp; tmp = tmp->next) 
	{
		tmp_set = tmp->cmd_set;
		
		if (!StrCaseCmp(argv[1], tmp_set->name))
		{
			printf("Available commands on the %s pipe:\n\n", tmp_set->name);

			i = 0;
			tmp_set++;
			while(tmp_set->name) {
				printf("%20s", tmp_set->name);
                                tmp_set++;
				i++;
				if (i%4 == 0)
					printf("\n");
			}
			
			/* drop out of the loop */
			break;
		}
        }
	printf("\n\n");

	return NT_STATUS_OK;
}

/* Display help on commands */

static NTSTATUS cmd_help(struct cli_state *cli, TALLOC_CTX *mem_ctx,
                         int argc, const char **argv)
{
	struct cmd_list *tmp;
        struct cmd_set *tmp_set;

        /* Usage */

        if (argc > 2) {
                printf("Usage: %s [command]\n", argv[0]);
                return NT_STATUS_OK;
        }

        /* Help on one command */

        if (argc == 2) {
                for (tmp = cmd_list; tmp; tmp = tmp->next) {
                        
                        tmp_set = tmp->cmd_set;

                        while(tmp_set->name) {
                                if (strequal(argv[1], tmp_set->name)) {
                                        if (tmp_set->usage &&
                                            tmp_set->usage[0])
                                                printf("%s\n", tmp_set->usage);
                                        else
                                                printf("No help for %s\n", tmp_set->name);

                                        return NT_STATUS_OK;
                                }

                                tmp_set++;
                        }
                }

                printf("No such command: %s\n", argv[1]);
                return NT_STATUS_OK;
        }

        /* List all commands */

	for (tmp = cmd_list; tmp; tmp = tmp->next) {

		tmp_set = tmp->cmd_set;

		while(tmp_set->name) {

			printf("%15s\t\t%s\n", tmp_set->name,
			       tmp_set->description ? tmp_set->description:
			       "");

			tmp_set++;
		}
	}

	return NT_STATUS_OK;
}

/* Change the debug level */

static NTSTATUS cmd_debuglevel(struct cli_state *cli, TALLOC_CTX *mem_ctx,
                               int argc, const char **argv)
{
	if (argc > 2) {
		printf("Usage: %s [debuglevel]\n", argv[0]);
		return NT_STATUS_OK;
	}

	if (argc == 2) {
		DEBUGLEVEL = atoi(argv[1]);
	}

	printf("debuglevel is %d\n", DEBUGLEVEL);

	return NT_STATUS_OK;
}

static NTSTATUS cmd_quit(struct cli_state *cli, TALLOC_CTX *mem_ctx,
                         int argc, const char **argv)
{
	exit(0);
	return NT_STATUS_OK; /* NOTREACHED */
}

/* Built in rpcclient commands */

static struct cmd_set rpcclient_commands[] = {

	{ "GENERAL OPTIONS" },

	{ "help", 	cmd_help, 	  -1,	"Get help on commands", "[command]" },
	{ "?", 		cmd_help, 	  -1,	"Get help on commands", "[command]" },
	{ "debuglevel", cmd_debuglevel,   -1,	"Set debug level", "level" },
	{ "list",	cmd_listcommands, -1,	"List available commands on <pipe>", "pipe" },
	{ "exit", 	cmd_quit, 	  -1,	"Exit program", "" },
	{ "quit", 	cmd_quit, 	  -1,	"Exit program", "" },

	{ NULL }
};

static struct cmd_set separator_command[] = {
	{ "---------------", NULL,	-1,	"----------------------" },
	{ NULL }
};


/* Various pipe commands */

extern struct cmd_set lsarpc_commands[];
extern struct cmd_set samr_commands[];
extern struct cmd_set spoolss_commands[];
extern struct cmd_set netlogon_commands[];
extern struct cmd_set srvsvc_commands[];
extern struct cmd_set dfs_commands[];
extern struct cmd_set reg_commands[];
extern struct cmd_set ds_commands[];

static struct cmd_set *rpcclient_command_list[] = {
	rpcclient_commands,
	lsarpc_commands,
	ds_commands,
	samr_commands,
	spoolss_commands,
	netlogon_commands,
	srvsvc_commands,
	dfs_commands,
	reg_commands,
	NULL
};

static void add_command_set(struct cmd_set *cmd_set)
{
	struct cmd_list *entry;

	if (!(entry = (struct cmd_list *)malloc(sizeof(struct cmd_list)))) {
		DEBUG(0, ("out of memory\n"));
		return;
	}

	ZERO_STRUCTP(entry);

	entry->cmd_set = cmd_set;
	DLIST_ADD(cmd_list, entry);
}


/**
 * Call an rpcclient function, passing an argv array.
 *
 * @param cmd Command to run, as a single string.
 **/
static NTSTATUS do_cmd(struct cli_state *cli,
		       struct cmd_set *cmd_entry,
		       int argc, char **argv)
{
	NTSTATUS result;
	
	TALLOC_CTX *mem_ctx;

	/* Create mem_ctx */

	if (!(mem_ctx = talloc_init("do_cmd"))) {
		DEBUG(0, ("talloc_init() failed\n"));
		return NT_STATUS_UNSUCCESSFUL;
	}

	/* Open pipe */

	if (cmd_entry->pipe_idx != -1)
		if (!cli_nt_session_open(cli, cmd_entry->pipe_idx)) {
			DEBUG(0, ("Could not initialize pipe\n"));
			return NT_STATUS_UNSUCCESSFUL;
		}

	/* Run command */

	result = cmd_entry->fn(cli, mem_ctx, argc, (const char **) argv);

	/* Cleanup */

	if (cmd_entry->pipe_idx != -1)
		cli_nt_session_close(cli);

	talloc_destroy(mem_ctx);

	return result;
}


/**
 * Process a command entered at the prompt or as part of -c
 *
 * @returns The NTSTATUS from running the command.
 **/
static NTSTATUS process_cmd(struct cli_state *cli, char *cmd)
{
	struct cmd_list *temp_list;
	NTSTATUS result = NT_STATUS_OK;
	int ret;
	int argc;
	char **argv = NULL;

	if ((ret = poptParseArgvString(cmd, &argc, (const char ***) &argv)) != 0) {
		fprintf(stderr, "rpcclient: %s\n", poptStrerror(ret));
		return NT_STATUS_UNSUCCESSFUL;
	}


	/* Walk through a dlist of arrays of commands. */
	for (temp_list = cmd_list; temp_list; temp_list = temp_list->next) {
		struct cmd_set *temp_set = temp_list->cmd_set;

		while (temp_set->name) {
			if (strequal(argv[0], temp_set->name)) {
				if (!temp_set->fn) {
					fprintf (stderr, "Invalid command\n");
					goto out_free;
				}

				result = do_cmd(cli, temp_set, argc, argv);

				goto out_free;
			}
			temp_set++;
		}
	}

	if (argv[0]) {
		printf("command not found: %s\n", argv[0]);
	}

out_free:
	if (!NT_STATUS_IS_OK(result)) {
		printf("result was %s\n", nt_errstr(result));
	}

	if (argv) {
		/* NOTE: popt allocates the whole argv, including the
		 * strings, as a single block.  So a single free is
		 * enough to release it -- we don't free the
		 * individual strings.  rtfm. */
		free(argv);
	}
	
	return result;
}


/* Main function */

 int main(int argc, char *argv[])
{
	static int		got_pass = 0;
	BOOL 			interactive = True;
	int 			opt;
	static char		*cmdstr = "";
	const char *server;
	struct cli_state	*cli;
	fstring 		password="",
				username="",
		domain="";
	static char 		*opt_authfile=NULL,
				*opt_username=NULL,
				*opt_domain=NULL,
 	                        *opt_logfile=NULL,
	                        *opt_ipaddr=NULL;
	pstring 		logfile;
	struct cmd_set 		**cmd_set;
	struct in_addr 		server_ip;
	NTSTATUS 		nt_status;

	/* make sure the vars that get altered (4th field) are in
	   a fixed location or certain compilers complain */
	poptContext pc;
	struct poptOption long_options[] = {
		POPT_AUTOHELP
		{"authfile",	'A', POPT_ARG_STRING,	&opt_authfile, 'A', "File containing user credentials", "AUTHFILE"},
		{"nopass",	'N', POPT_ARG_NONE,	&got_pass, 'N', "Don't ask for a password"},
		{"user", 'U', POPT_ARG_STRING,	&opt_username, 'U', "Set the network username", "USER"},
		{"workgroup", 'W', POPT_ARG_STRING, 	&opt_domain, 'W', "Set the domain name for user account", "DOMAIN"},
		{"command",	'c', POPT_ARG_STRING,	&cmdstr, 'c', "Execute semicolon separated cmds", "COMMANDS"},
		{"logfile",	'l', POPT_ARG_STRING,	&opt_logfile, 'l', "Logfile to use instead of stdout", "LOGFILE" },
		{"dest-ip", 'I', POPT_ARG_STRING,   &opt_ipaddr, 'I', "Specify destination IP address", "IP"},
		{ NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_debug },
		{ NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_configfile },
		{ NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_version},
		{ NULL }
	};

	setlinebuf(stdout);

	/* Parse options */

	pc = poptGetContext("rpcclient", argc, (const char **) argv,
			    long_options, 0);

	if (argc == 1) {
		poptPrintHelp(pc, stderr, 0);
		return 0;
	}
	
	while((opt = poptGetNextOpt(pc)) != -1) {
		switch (opt) {
		case 'A':
			/* only get the username, password, and domain from the file */
			read_authfile (opt_authfile, username, password, domain);
			if (strlen (password))
				got_pass = 1;
			break;
			
		case 'l':
			slprintf(logfile, sizeof(logfile) - 1, "%s.client", 
				 opt_logfile);
			lp_set_logfile(logfile);
			interactive = False;
			break;
			
		case 'U': {
			char *lp;

			fstrcpy(username,opt_username);

			if ((lp=strchr_m(username,'%'))) {
				*lp = 0;
				fstrcpy(password,lp+1);
				got_pass = 1;
				memset(strchr_m(opt_username,'%') + 1, 'X',
				       strlen(password));
			}
			break;
		}
		case 'I':
		        if ( (server_ip.s_addr=inet_addr(opt_ipaddr)) == INADDR_NONE ) {
				fprintf(stderr, "%s not a valid IP address\n",
					opt_ipaddr);
				return 1;
			}
		case 'W':
			fstrcpy(domain, opt_domain);
			break;
		}
	}

	/* Get server as remaining unparsed argument.  Print usage if more
	   than one unparsed argument is present. */

	server = poptGetArg(pc);
	
	if (!server || poptGetArg(pc)) {
		poptPrintHelp(pc, stderr, 0);
		return 1;
	}

	poptFreeContext(pc);

	/* the following functions are part of the Samba debugging
	   facilities.  See lib/debug.c */
	setup_logging("rpcclient", interactive);
	if (!interactive) 
		reopen_logs();
	
	/* Load smb.conf file */

	if (!lp_load(dyn_CONFIGFILE,True,False,False))
		fprintf(stderr, "Can't load %s\n", dyn_CONFIGFILE);

	load_interfaces();

	if (!init_names())
		return 1;

	/* Resolve the IP address */

	if (!opt_ipaddr && !resolve_name(server, &server_ip, 0x20))  {
		fprintf(stderr, "Unable to resolve %s\n", server);
		return 1;
	}
	
	/*
	 * Get password
	 * from stdin if necessary
	 */

	if (!got_pass) {
		char *pass = getpass("Password:");
		if (pass) {
			fstrcpy(password, pass);
		}
	}
	
	if (!strlen(username) && !got_pass)
		get_username(username);
		
	nt_status = cli_full_connection(&cli, lp_netbios_name(), server, 
					&server_ip, 0,
					"IPC$", "IPC",  
					username, domain,
					password, 0, NULL);
	
	if (!NT_STATUS_IS_OK(nt_status)) {
		DEBUG(0,("Cannot connect to server.  Error was %s\n", nt_errstr(nt_status)));
		return 1;
	}

	memset(password,'X',sizeof(password));

	/* Load command lists */

	cmd_set = rpcclient_command_list;

	while(*cmd_set) {
		add_command_set(*cmd_set);
		add_command_set(separator_command);
		cmd_set++;
	}

	fetch_machine_sid(cli);
 
       /* Do anything specified with -c */
        if (cmdstr[0]) {
                char    *cmd;
                char    *p = cmdstr;
 
                while((cmd=next_command(&p)) != NULL) {
                        process_cmd(cli, cmd);
                }
		
		cli_shutdown(cli);
                return 0;
        }

	/* Loop around accepting commands */

	while(1) {
		pstring prompt;
		char *line;

		slprintf(prompt, sizeof(prompt) - 1, "rpcclient $> ");

		line = smb_readline(prompt, NULL, completion_fn);

		if (line == NULL)
			break;

		if (line[0] != '\n')
			process_cmd(cli, line);
	}
	
	cli_shutdown(cli);
	return 0;
}
